internal/terraform/node_resource_plan_instance.go

```diff
@@ -417,6 +423,17 @@ func (n *NodePlannableResourceInstance) managedResourceExecute(ctx EvalContext)
 			checkRuleSeverity,
 		)
 		diags = diags.Append(checkDiags)
+
+		if refreshDeferred != nil {
+			ctx.Deferrals().ReportResourceInstanceDeferred(addr, deferred.Reason, &plans.ResourceInstanceChange{
+				Addr: n.Addr,
+				Change: plans.Change{
+					Action: plans.Read,
+					Before: instanceRefreshState.Value,
```

Comment:
I think if the refresh happens, `instanceRefreshState` isn't the before value but it should be the after value. At line 216 we change the `instanceRefreshState` to the response from the refresh command which will be the refresh's best guess as to the after (post-refresh) value.

You might need to track a `postRefreshState` value that is used here, and then use `instanceRefreshState` as the after.